examples/vault-ec2-auth/user-data-auth-client.sh

```diff
@@ -0,0 +1,110 @@
+#!/bin/bash
+# This script is meant to be run in the User Data of each EC2 Instance while it's booting. The script uses the
+# run-consul script to configure and start Consul in client mode. Note that this script assumes it's running in an AMI
+# built from the Packer template in examples/vault-consul-ami/vault-consul.json.
+
+set -e
+
+# Send the log output from this script to user-data.log, syslog, and the console
+# From: https://alestic.com/2010/12/ec2-user-data-output/
+exec > >(tee /var/log/user-data.log|logger -t user-data -s 2>/dev/console) 2>&1
+
+# These variables are passed in via Terraform template interpolation
+/opt/consul/bin/run-consul --client --cluster-tag-key "${consul_cluster_tag_key}" --cluster-tag-value "${consul_cluster_tag_value}"
+
+# Retrieves the pkcs7 certificate from instance metadata
+# The vault role name is filled by terraform
+# The role itself is created when configuting the vault cluster
+pkcs7=$(curl -s http://169.254.169.254/latest/dynamic/instance-identity/pkcs7 | tr -d '\n')
+data=$(cat <<EOF
+{
+  "role": "${example_role_name}",
+  "pkcs7": "$pkcs7"
+}
+EOF
+)
+
+# run-consul is running on the background, so we have to wait for it and
+# we also have to for wait for vault server to be booted and unsealed before it can accept this request
+# so in case this fails we retry.
+# The boolean operations with the exit status are there to temporarily circumvent the "set -e" at the
+# beginning of this script which exits the script immediatelly for error status.
+for i in $(seq 1 30); do
+  login_output=$(curl --request POST --data "$data" "https://vault.service.consul:8200/v1/auth/aws/login") && exit_status=0 || exit_status=$?
+  if [[ $exit_status -eq 0 ]]; then break; fi
+  sleep 10
+done; (exit $exit_status)
```

Comment:
This is hard to read, and it's not obvious to me why this wouldn't always exit the script. I think a better approach is to wrap this code in a function (e.g., `login_to_vault`) that calls `return` if login is successful or `exit $exit_status` if, after N retries, login was still unsuccessful.